### PR TITLE
Bootstrapping a React Project

### PR DESCRIPTION
**Pull request by unicodeveloper**
first-screenshot
new

### CODE REVIEW

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+related:
+- 2015-04-09-adding-authentication-to-your-react-flux-app
+- 2015-08-25-logging-and-debugging-in-react-with-flux-replaying-your-users-actions/
+- 2016-01-04-secure-your-react-and-redux-app-with-jwt-authentication
+---
+
+---
+
+**TL;DR:** The Javascript ecosystem has a plethora of libraries and frameworks for front-end development. ReactJS is one of the young and shiny new library on the block. Just a little over a year old, it has gained lots of traction from javascript developers around the world. One of ReactJS's selling points is the ability to easily build reusable components. Another salient point is that not only does it perform on the client side, it can also be rendered on the server side. In fact, when compared to other front-end libraries and frameworks currently available, ReactJS claims to be the "V" in "MVC".
+
```
Comment (gnandretta) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: I think they no longer claim that React is the V in MVC. They removed that statement from their website and started giving talks about "rethinking MVC".

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+---
+
+## Why React?
+**ReactJS** was built by Facebook to solve one problem: building large applications with data that changes over time. It utilizes the concept of the Virtual DOM that selectively renders subtrees of nodes based upon state changes. The component based architecture that **ReactJs** emulates makes it incredible to work with. Components in **ReactJS** are self-contained views that get their data from triggered events or as inherited property from other components.
+
```
Comment (gnandretta) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: I had a hard time wrapping my head around this. The key takeaway is that you don't need to make changes to the DOM, react can make them for you and for that it uses a virtual dom.

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown (hunk 1 above)
Comment (unicodeveloper) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Ohh...then, I'll have to remove that

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown (hunk 2 above)
Comment (unicodeveloper) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Yeah...I made mention of React making use of the Virtual Dom. What part was difficult to wrap your head around?

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+  url: https://twitter.com/unicodeveloper?lang=en
+  avatar: https://en.gravatar.com/avatar/1097492785caf9ffeebffeb624202d8f?s=200
+  mail: prosper.otemuyiwa@auth0.com
+design:
+  bg_color: "rgb(25, 25, 25)"
+  bg_merge: true
+  image: https://cdn.auth0.com/blog/react-js/react.png
+  image_size: "80%"
+  image_bg_color: "rgb(25, 25, 25)"
+tags:
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: I believe we were limiting the tags to 3-4 most relevant ones. I would double check with Tony though to make sure.

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+- material-ui
+related:
+- 2015-04-09-adding-authentication-to-your-react-flux-app
+- 2015-08-25-logging-and-debugging-in-react-with-flux-replaying-your-users-actions/
+- 2016-01-04-secure-your-react-and-redux-app-with-jwt-authentication
+---
+
+---
+
+**TL;DR:** The Javascript ecosystem has a plethora of libraries and frameworks for front-end development. ReactJS is one of the young and shiny new library on the block. Just a little over a year old, it has gained lots of traction from javascript developers around the world. One of ReactJS's selling points is the ability to easily build reusable components. Another salient point is that not only does it perform on the client side, it can also be rendered on the server side. In fact, when compared to other front-end libraries and frameworks currently available, ReactJS claims to be the "V" in "MVC".
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: library => libraries.

Also React is ~ 3 years old now - but you can mention something like: even though it's only a few years old, it has gained lots of traction..

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+      <div>
+        The easiest way to add authentication to your app is via Auth0!
+      </div>
+    );
+  }
+}
+
+```
+
+## Let's get started
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Let's Get Started

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+
+Bootstrapping a **ReactJS** project involves setting up lots of tools, and it can be really be a daunting task. Thankfully there are several tools provided for developers by the community to aid in setting up your project without breaking a sweat. In this post, we'll cover tools that you should be aware of in bootstrapping your next **ReactJS** project.
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: be really be => really be

you have an extra be :)

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown (hunk 4 above)
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Javascript => JavaScript

Just so we're consistent.

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+
+## Build Tools
+
+The first step is to set up a development environment for your **ReactJS** project. [Webpack](http://webpack.github.io/), a bundler utility and [Babel](http://babeljs.io/) are key tools here. Babel lets us write code that uses new ES6 features, and then transpiles that code into standard ES5 code that can run in older JavaScript environments. When setting up a new project:
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: ...that can run in older JavaScript environments...  

Something just feels off about this sentence. Currently none of the browsers really support the entire ES6 spec, so transpilation is needed for almost every browser. Maybe you can reword this to say something along the lines of "then transpiles that code into standard ES5 code that can actually run in the browser." WDYT?

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+
+  * I assume Nodejs and NPM is installed, run this command:  `npm init` to create a `package.json` file
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Nodejs => Node.js

Maybe link to nodejs.org?

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown (hunk 7 above)
Comment (unicodeveloper) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Chrome Canary actually supports about 99%of ES6 and ES7

*Best Regards,*

*Otemuyiwa Prosper | IT Trainer | Andela  andela.co <http://andela.co>*
(c) +234-708-974-0924 | @unicodeveloper <https://twitter.com/unicodeveloper> |
prosper.otemuyiwa1 ( *Skype* )

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+    "start": "node_modules/.bin/webpack-dev-server --progress --inline --hot",
+  }
+}
+```
+
+The code above shows we have added `--hot` option. This simply enables the the hot reloading once you start your app by running `npm run start`.
+
+## Routing
+
+Routing is a very essential part of any application. The most popular choice for doing this in a **ReactJS** application is [React Router](https://github.com/rackt/react-router). In fact, a lot of developers tip it as the official routing system for **ReactJS**. React Router is dependent on the [history](https://github.com/ReactJSTraining/history) module. When you want to deal with routing, you have to install them side by side in your project as shown below:
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: you have to install them side by side in your project as show below: => you have to install the React Router module in your project as show below:

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+      <IndexRoute component={Home} />
+      <Route path='about' component={About} />
+      <Route path='features' component={Features} />
+    </Route>
+  </Router>,
+  document.getElementById('app')
+)
+```
+
+![Routing](https://cdn.auth0.com/blog/react:routing.png)
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: is the ":" a valid item in a url. when I access this url directly it downloads the image to my desktop instead of displaying it in the browser?

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+
+You actually might not need React Router. More info [here](https://medium.com/@tarkus/you-might-not-need-react-router-38673620f3d#.u64h0gcl4)
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: I would expand on this just a little bit. Why would I not need React Router?

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+
+## Internationalization
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Internationalization (I18N)

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+
+There are over 6 billion people in the world. These people speak different languages. One way of making your app available to billions of people around the world is providing support for native languages on your app. In a **ReactJS** project, you can easily achieve that with a very good and well tested library [react-intl](https://github.com/yahoo/react-intl). By default, [react-intl](https://github.com/yahoo/react-intl) ships with the locale data for English built-in to the library's runtime. When you need to format data in another locale, include its data. This library provides React components and an API to format dates, numbers, and strings, including pluralization and handling translations. Oh, it supports over 150 languages!
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: on your app => in your app

Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: "When you need to format data in another locale, include its data." Is it possible to expand on this a little bit?

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 14]
```diff
@@ -0,0 +1,442 @@
...
+  <App />,
+  document.getElementById('app')
+);
+```
+
+One effective way of bootstrapping your **ReactJS** project is to start by designing your UI components and then glue them together, that way you can split up the initial setup effort into several small parts along the project lifecycle. [Pure UI](http://rauchg.com/2015/pure-ui/) explains this in detail. I also recommend these tools: [Carte-blanche](https://github.com/carteb/carte-blanche), [React Storybook](https://github.com/kadirahq/react-storybook), [uiharness.com](http://www.uiharness.com/start). They will help a lot!
+
+## Network Requests
+
+In a situation where you have to fetch data from an external API e.g Github into your **ReactJS** application, there are several tools you can use to achieve that. I highly recommend [axios](https://github.com/mzabriskie/axios) and [superagent](https://github.com/visionmedia/superagent).
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: an external API e.g. Github => an external API, e.g. calling the GitHub API in your

Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: you can also remove "to achieve that" as it's redundant :)

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 15]
```diff
@@ -0,0 +1,442 @@
...
+
+_HTTP Request Example With Axios_
+
+```js
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: All code samples so far have been written in ES6 - any particular reason this one isn't? I think if we can keep them consistent it would be best for the reader.

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 16]
```diff
@@ -0,0 +1,442 @@
...
+
+![Data Flow](https://facebook.github.io/flux/img/flux-simple-f8-diagram-1300w.png)
+*(Source: [Facebook](https://facebook.github.io/flux/docs/overview.html))*
+
+
+### Redux
+Redux evolves the idea of Flux but avoid its complexity. It's a state management library with minimal API but completely predictable behavior, so it is possible to implement logging, hot reloading, time travel, universal apps, record and replay, without any buy-in from the developer. You can also install it via NPM like so: `npm install redux redux-devtools --save`. Redux attempts to make state mutations predictable by imposing certain restrictions on how and when updates can happen. Redux has three fundamental principles:
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: with minimal API => with a minimal API

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 17]
```diff
@@ -0,0 +1,442 @@
...
+
+* Single source of truth
+* State is read-only
+* Changes are made with pure functions
+
+
+Read More about Redux [here](http://redux.js.org/). [Here](Awesome list of Redux examples and middlewares) is also an awesome list of Redux examples and middlewares. Another alternative for state management within your **ReactJS** application is [Alt](http://alt.js.org/). More information about Alt [here](http://alt.js.org/guide/).
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Read More => Read more

Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: [Here] is missing a link :)

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 18]
```diff
@@ -0,0 +1,442 @@
...
+
+## Authentication
+
+Authentication is an important part of any application. And the best way to do user authentication for single page apps is via JSON Web Tokens(JWT). A typical authentication flow is this:
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: And the => The

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 19]
```diff
@@ -0,0 +1,442 @@
...
+
+### Implementing Authentication with Auth0
+
+**Auth0** issues [JSON Web Tokens](https://jwt.io/) on every login for your users. This means that you can have a solid [identity infrastructure](https://auth0.com/docs/identityproviders), including [single sign-on](https://auth0.com/docs/sso/single-sign-on), user management, support for social identity providers (Facebook, Github, Twitter, etc.), enterprise identity providers (Active Directory, LDAP, SAML, etc.) and your own database of users with just a few lines of code. Multifactor Authentication, Single sign-on and passwordless-login is also a breeze with **Auth0**.
+
+With Auth0, you can add authentication to any app in under 10 minutes and implement features like social login, multifactor auth, and single sign-on at the flip of a switch. It is the easiest way to add authentication to your app!
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: I don't know if we should repeat the features again - this is the third time in 2 paragraphs that we mention social login, sso, etc. Maybe rework this section a bit. WDYT?

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown (hunk 3 above)
Comment (unicodeveloper) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: I'll check with Tony on this

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 20]
```diff
@@ -0,0 +1,442 @@
...
+
+Without a backend, you can persist data in your Single Page App by using Firebase. In a **Reactjs** app, all you simply need is [ReactFire](https://github.com/firebase/reactfire). It is a **ReactJS** mixin for easy Firebase integration. With ReactFire, it only takes a few lines of JavaScript to integrate Firebase data into React apps via the ReactFireMixin.
+
+```bash
+npm install reactfire react firebase --save
+```
+
+_TodoList Example_
+
+```js
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Again if we have a version with ES6 code it may be better for reader.

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown (hunk 4 above)
Comment (unicodeveloper) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: 👍

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown (hunk 10 above)
Comment (unicodeveloper) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: All the other images uploaded via the new upload slack bot are downloaded to the desktop instead of been displayed in the browser. That's what I have noticed so far

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown (hunk 15 above)
Comment (unicodeveloper) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: 😬 I'll rewrite this in ES6

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 21]
```diff
@@ -0,0 +1,442 @@
...
+
+
+## Conclusion
+
+There are several tools that will help bootstrap your react app, we selected a couple that we consider are quite good to have your application up and running in no time. But feel free to search for your own tools, and if you think that we are missing something, let us know in the comments. Setting up a React Project should be painless!
+
+{% include tweet_quote.html quote_text="Setting up a React Project should be painless!" %}
```
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: React Project => React project

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown (hunk 19 above)
Comment (unicodeveloper) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: I thought mentioning it again might be good for SEO. Same way I was always mentioning ReactJS

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown (hunk 10 above)
Comment (kukicado) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Ah ok - I'll report that as a bug to the right crew then :) it looks like they work fine on the actual blog so I guess it's not a big deal for now.

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 22]
```diff
@@ -0,0 +1,442 @@
...
+
+Bootstrapping a **ReactJS** project involves setting up lots of tools, and it can be really be a daunting task. Thankfully there are several tools provided for developers by the community to aid in setting up your project without breaking a sweat. In this post, we'll cover tools that you should be aware of in bootstrapping your next **ReactJS** project.
+
+## Build Tools
+
+The first step is to set up a development environment for your **ReactJS** project. [Webpack](http://webpack.github.io/), a bundler utility and [Babel](http://babeljs.io/) are key tools here. Babel lets us write code that uses new ES6 features, and then transpiles that code into standard ES5 code that can run in older JavaScript environments. When setting up a new project:
+
+  * I assume Nodejs and NPM is installed, run this command:  `npm init` to create a `package.json` file
+  * Run this command in your terminal:  `npm install react react-dom babel-core babel-loader babel-preset-es2015 babel-preset-react webpack webpack-dev-server --save`
+  * Create a `webpack.config.js` file. This file handles bundling all our assets, converting JSX to JS files and launching the development server.
```
Comment (gnandretta) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: A sentence to define what bundle means will also help here.

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 23]
```diff
@@ -0,0 +1,442 @@
...
+  },
+  plugins: debug ? [] : [
+    new webpack.optimize.DedupePlugin(),
+    new webpack.optimize.OccurenceOrderPlugin(),
+    new webpack.optimize.UglifyJsPlugin({ mangle: false, sourcemap: false }),
+  ],
+};
+```
+
+The application's entry point is set at the `entry`, Webpack will bundle all the `JS` and `JSX` files into the file that's specified in the output object. Webpack dev server is set to inline which allows a kind of reload on the fly and the application will be served at port `3333`. In the module object, we specify the babel loader to use `react` and `es2015` presets for converting `ES6` and `React` code into code that the browser understands. The webpack plugins add the ability to use class properties and decorators.
```
Comment (gnandretta) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Maybe explain what a preset is

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 24]
```diff
@@ -0,0 +1,442 @@
...
+```
+
+The code above shows we have added `--hot` option. This simply enables the the hot reloading once you start your app by running `npm run start`.
+
+## Routing
+
+Routing is a very essential part of any application. The most popular choice for doing this in a **ReactJS** application is [React Router](https://github.com/rackt/react-router). In fact, a lot of developers tip it as the official routing system for **ReactJS**. React Router is dependent on the [history](https://github.com/ReactJSTraining/history) module. When you want to deal with routing, you have to install them side by side in your project as shown below:
+
+```bash
+npm install --save react-router history
```
Comment (gnandretta) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: react-router already depends on history (see https://github.com/reactjs/react-router/blob/112cce48255f930df3270382665223cd41b6f1fa/package.json#L35)  so it should be installed automatically when you install react-router. Is there is a reason to install history explicitly? If so it might be good to include in the post.

Comment (unicodeveloper) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Thanks for pointing that out. All the tutorials I followed stated it should be installed alongside react-router. Since it's already installed automatically, there is actually no reason to install explicitly. I'll clear that now

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown (hunk 16 above)
Comment (gnandretta) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Redux doesn't avoid the complexity of flux. You can argue that redux is more complex than flux. While flux suggests using various stores, redux uses only one. That is what makes it easier to have a time-travel debugger. The point of redux is that you most of the cool stuff listed here for free and it also has a vibrant ecosystem around it (so there are a lots of tools).

Comment (unicodeveloper) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Thanks for the explanation. I'm loving this 👍

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 25]
```diff
@@ -0,0 +1,442 @@
...
+});
+
+ReactDOM.render(<TodoApp />, document.getElementById('todoApp'));
+```
+
+More information about persisting your data using ReactFire [here](https://github.com/firebase/reactfire/blob/master/docs/quickstart.md).
+
+## Testing
+
+Most projects become a mountain of spaghetti code at some point during development due to lack of solid tests or no tests at all. **ReactJS** apps are no different, and can be tested simply if you know some core principles. When writing tests for **ReactJS** code, it is helpful to pull out any functionality that doesn't have to do with any UI components into separate modules, so that they can be tested separately. Tools for unit testing those functionalities are [mocha](https://github.com/mochajs/mocha), [expect](https://github.com/mjackson/expect), [chai](https://github.com/chaijs/chai), [jasmine](https://github.com/jasmine/jasmine).
```
Comment (gnandretta) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: IMO test help code to turn into spaghetti because "the tests pass, so although I don't understand the change I made it should work" hahaha

Comment (unicodeveloper) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: loool

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 26]
```diff
@@ -0,0 +1,442 @@
...
+
+### Implementing Authentication with Auth0
+
+**Auth0** issues [JSON Web Tokens](https://jwt.io/) on every login for your users. This means that you can have a solid [identity infrastructure](https://auth0.com/docs/identityproviders), including [single sign-on](https://auth0.com/docs/sso/single-sign-on), user management, support for social identity providers (Facebook, Github, Twitter, etc.), enterprise identity providers (Active Directory, LDAP, SAML, etc.) and your own database of users with just a few lines of code. Multifactor Authentication, Single sign-on and passwordless-login is also a breeze with **Auth0**.
+
+With Auth0, you can add authentication to any app in under 10 minutes and implement features like social login, multifactor auth, and single sign-on at the flip of a switch. It is the easiest way to add authentication to your app!
+
+A full implementation of Authentication with Auth0 in a **ReactJS** application is [here](https://davidwalsh.name/react-authentication).
```
Comment (diegopoza) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Well it won't affect our SEO, but I think it will be better to link to our quickstart. (https://auth0.com/docs/quickstart/spa/react)

Comment (unicodeveloper) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Alright, let me link to it !

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 27]
```diff
@@ -0,0 +1,470 @@
...
+- enzyme
+related:
+- 2015-04-09-adding-authentication-to-your-react-flux-app
+- 2015-08-25-logging-and-debugging-in-react-with-flux-replaying-your-users-actions/
+- 2016-01-04-secure-your-react-and-redux-app-with-jwt-authentication
+---
+
+---
+
+**TL;DR:** The JavaScript ecosystem has a plethora of libraries and frameworks for front-end development. ReactJS is one of the young and shiny new library on the block. Even though it's only a few years old, it has gained lots of traction from JavaScript developers around the world. One of ReactJS's selling points is the ability to easily build reusable components. Another salient point is that not only does it perform on the client side, it can also be rendered on the server side.
```
Comment (DylanMeeus) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Consider changing:

> reactJS is one of the young and shiny new library on the block

into:

> reactJS is one of the young and shiny new libraries on the block

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 28]
```diff
@@ -0,0 +1,470 @@
...
+
+---
+
+## Why React?
+**ReactJS** was built by Facebook to solve one problem: building large applications with data that changes over time. It utilizes the concept of the Virtual DOM that selectively renders subtrees of nodes based upon state changes. The component based architecture that **ReactJs** emulates makes it incredible to work with. Components in **ReactJS** are self-contained views that get their data from triggered events or as inherited property from other components. There are basically two main ways of writing components. Props and States in **ReactJs** also makes it fun to work with. Learn more about the basic concepts of React [here](https://scotch.io/tutorials/learning-react-getting-started-and-concepts).
```
Comment (DylanMeeus) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: >  There are basically two main ways of writing components. Props and States in **ReactJs** also makes it fun to work with.

That does not read 'naturally' to me. You speak of two main ways of writing components, maybe shortly list them?  Don't jump to 'props and states' [..] fun to work with.' 

I would rewrite this more or less like this: 

> There are essentially two main ways of writing components, [x and y].

Just a suggestion, I also replaced 'basically' with 'essentially' which is really nitpicking. (Sorry for that).

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 29]
```diff
@@ -0,0 +1,470 @@
...
+
+## Building a React App
+
+The first step is to set up a development environment for your **ReactJS** project. [Webpack](http://webpack.github.io/), a bundler utility and [Babel](http://babeljs.io/) are key tools here. Babel lets us write code that uses new ES6 features, and then transpiles that code into standard ES5 code that can actually run in the browser. When setting up a new project:
+
+  * I assume [Node.js](https://nodejs.org) and NPM is installed, run this command:  `npm init` to create a `package.json` file
+  * Run this command in your terminal:  `npm install react react-dom babel-core babel-loader babel-preset-es2015 babel-preset-react webpack webpack-dev-server --save`
+  * Create a `webpack.config.js` file. This file handles bundling all our assets, converting JSX to JS files and launching the development server. Bundling refers to combining, compiling and minimizing multiple files into just one file to be served to the client.
+
+**Note:** JSX is a JavaScript syntax extension that looks similar to XML. You can write concise HTML/XML-like structures in the same file as you write JavaScript code. With Babel, JSX files can be transpiled into actual javascript code.
```
Comment (DylanMeeus) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Consider adding a link to the specification? (I noticed that at the moment it is a draft) https://facebook.github.io/jsx/

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 30]
```diff
@@ -0,0 +1,470 @@
...
+    "start": "node_modules/.bin/webpack-dev-server --progress --inline --hot",
+  }
+}
+```
+
+The code above shows we have added `--hot` option. This simply enables the the hot reloading once you start your app by running `npm run start`.
+
+## Routing
+
+Routing is a very essential part of any application. The most popular choice for doing this in a **ReactJS** application is [React Router](https://github.com/rackt/react-router). In fact, a lot of developers tip it as the official routing system for **ReactJS**. When you want to deal with routing, you have to install the React Router module in your project as shown below:
```
Comment (DylanMeeus) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: > Routing is an essential part

The word 'very' does not add anything here. (Which is quite frequently the case). 

> a lot of developers tip it as ..

'tip it as' is an odd expression to me. This might just be the case for me, so you can ignore this comment. But I would change that to something like:

> a lot of developers consider it the official routing ..

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 31]
```diff
@@ -0,0 +1,470 @@
...
+)
+```
+
+![Routing](https://cdn.auth0.com/blog/react:routing.png)
+
+You actually might not need React Router. More info [here](https://medium.com/@tarkus/you-might-not-need-react-router-38673620f3d#.u64h0gcl4)
+
+## Internationalization (I18N)
+
+There are over 6 billion people in the world. These people speak different languages. One way of making your app available to billions of people around the world is providing support for native languages in your app. In a **ReactJS** project, you can easily achieve that with a very good and well tested library [react-intl](https://github.com/yahoo/react-intl). By default, [react-intl](https://github.com/yahoo/react-intl) ships with the locale data for English built-in to the library's runtime. This library provides React components and an API to format dates, numbers, and strings, including pluralization and handling translations. Oh, it supports over 150 languages!
```
Comment (DylanMeeus) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: I'm sorry for nitpicking again. 

> There are over 6 billion people in the world.

That is true, the estimate is actually 7.4 billion. Thus I'd say

> There are over 7 billion people in the world.

The last sentence (perhaps even without the 'Oh' but I get why it is there.

> Oh, and it supports over 150 languages!

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 32]
```diff
@@ -0,0 +1,470 @@
...
+
+Flux is an application architecture for React that utilizes a unidirectional flow for building client-side web applications. With Flux, when a user interacts with a React view, the view propagates an action through a central dispatcher, to the various stores that hold the application's data and business logic, which updates all of the views that are affected. When choosing a dispatcher for your app, Facebook's [dispatcher](https://github.com/facebook/flux/blob/master/src/Dispatcher.js) library should come in handy. It's easy to instantiate and use. Alongside this library, you will need any good Javascript event library. NodeJS [EventEmmiter](https://nodejs.org/api/events.html#events_class_events_eventemitter) module is a good option. You can install flux from [npm](https://www.npmjs.com), the dispatcher will be immediately available via `var Dispatcher = require('flux').Dispatcher;`. More details about the Flux pattern can be found [here](http://facebook.github.io/flux/docs/overview.html).
+
+![Data Flow](https://facebook.github.io/flux/img/flux-simple-f8-diagram-1300w.png)
+*(Source: [Facebook](https://facebook.github.io/flux/docs/overview.html))*
+
+
+### Redux
```
Comment (DylanMeeus) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: While it may be beyond the scope of this post, you have added links to various useful tools. For redux, the redux dev tools could come in quite handy. https://github.com/zalmoxisus/redux-devtools-extension

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 33]
```diff
@@ -0,0 +1,470 @@
...
+
+## Generators and Boilerplates
+
+A lot of tools have been mentioned in this post in relation to setting up different parts of a **ReactJS** app. If you don't intend writing your app from scratch, there are lots of generators and boilerplates that tie all these tools together to give you a great starting point for your app. One fantastic one is [React Starter Kit](https://github.com/kriasoft/react-starter-kit). It has a [Yeoman generator](https://www.npmjs.com/package/generator-react-fullstack). It's an isomorphic web app boilerplate that contains almost everything you need to build a **ReactJS** app. Another boilerplate is [React Static boilerplate](https://github.com/kriasoft/react-static-boilerplate), it helps you build a web app that can be hosted directly from CDNs like Firebase and Github Pages. Other alternatives are [React redux starter kit](https://github.com/davezuko/react-redux-starter-kit) and [React webpack generator](https://github.com/newtriks/generator-react-webpack).
```
Comment (DylanMeeus) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: > If you don't intend on writing your app

Maybe 

> One fantastic example is

Instead of

> One fantastic one is 

The repetition of the word 'One' looks a bit odd.

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown [hunk 34]
```diff
@@ -0,0 +1,470 @@
...
+
+
+## Conclusion
+
+There are several tools that will help bootstrap your React app, we looked at a couple that we consider are quite good, that will have your application up and running in no time. But feel free to search for your own tools, and if you think that we are missing something, let us know in the comments. Setting up a React project should be painless!
```
Comment (DylanMeeus) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: > That we consider quite good and will have your application up and running in no time.

**Comment by DylanMeeus** [comment]
Nice post 👍 
I have added some small comments related to spelling/grammar or just some suggestions

File: _posts/2016-07-11-bootstrapping-a-react-project.markdown (hunk 28 above)
Comment (unicodeveloper) [inline _posts/2016-07-11-bootstrapping-a-react-project.markdown]: Makes a lot of sense.
